Automattic/wp-calypso
# Editor: Account for custom post types in "New Post" type label

Related: #3695

This pull request seeks to reimplement `<EditorPostType />` using applicable post type labels from the REST API for custom post types. 

![EditorPostType](https://cloud.githubusercontent.com/assets/1779930/13512354/114224fc-e168-11e5-8129-9ece3dae14b5.png)

__Implementation notes:__

- Dispatches to and consumes from Redux edited post state
- Removes `post-editor/post-type-switcher` feature flag and functionality, as this has been disabled even in development environments since the launch of the editor
- Fixes incorrect state tree traversal for merging edited post (bug in #3183, specifically 2eeec8e895f2f6409c9e7e25b0fa02e41acfb5a8)

__Testing instructions:__

There should be no visual changes in this pull request. Ensure that "New Post", "Post", "New Page", and "Page" display correctly depending on the editor context.

You can observe the effect this will have in accounting for custom post types by applying the following patch to `<EditorPostType />` (to remove hardcoding) and running Calypso via `DISABLE_FEATURES=persist-redux make run`:

```diff
diff --git a/client/post-editor/editor-post-type/index.jsx b/client/post-editor/editor-post-type/index.jsx
index 8397a61..5520962 100644
--- a/client/post-editor/editor-post-type/index.jsx
+++ b/client/post-editor/editor-post-type/index.jsx
@@ -19,8 +19,6 @@ import QueryPostTypes from 'components/data/query-post-types';
 function EditorPostType( { siteId, isNew, typeSlug, type } ) {
 	let label;
 	switch ( typeSlug ) {
-		case 'page': label = isNew ? translate( 'New Page' ) : translate( 'Page', { context: 'noun' } ); break;
-		case 'post': label = isNew ? translate( 'New Post' ) : translate( 'Post', { context: 'noun' } ); break;
 		default: label = isNew ? get( type, 'labels.new_item' ) : get( type, 'labels.singular_name' );
 	}
 
@@ -30,7 +28,7 @@ function EditorPostType( { siteId, isNew, typeSlug, type } ) {
 
 	return (
 		<span className={ classes }>
-			{ siteId && 'page' !== typeSlug && 'post' !== typeSlug && (
+			{ siteId && (
 				<QueryPostTypes siteId={ siteId } />
 			) }
 			{ label || translate( 'Loading…' ) }

```

1. Navigate to the [Calypso post editor](http://calypso.localhost:3000/post)
2. Select a site
3. Observe that a placeholder label is shown where "New Post" would appear until the post types have loaded
4. Note that "New Post" is shown after the post type has loaded
